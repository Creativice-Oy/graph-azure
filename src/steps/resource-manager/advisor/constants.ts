import { RelationshipClass } from '@jupiterone/integration-sdk-core';
import { ANY_SCOPE } from '../constants';

export const AdvisorSteps = {
  RECOMMENDATIONS: 'rm-advisor-recommendations',
};

export const AdvisorEntities = {
  RECOMMENDATION: {
    _type: 'azure_advisor_recommendation',
    _class: ['Finding'],
    resourceName: '[RM] Advisor Recommendation',
  },
};

export const AdvisorRelationships = {
  ANY_RESOURCE_HAS_FINDING: {
    _type: 'ANY_SCOPE_has_finding',
    sourceType: ANY_SCOPE,
    _class: RelationshipClass.HAS,
    targetType: AdvisorEntities.RECOMMENDATION._type,
  },
};
