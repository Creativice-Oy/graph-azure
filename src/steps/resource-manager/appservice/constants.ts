import { createResourceGroupResourceRelationshipMetadata } from '../utils/createResourceGroupResourceRelationship';

export const AppServiceSteps = {
  APPS: 'rm-appservice-apps',
};

export const AppServiceEntities = {
  WEB_APP: {
    _type: 'azure_web_app',
    _class: ['Application'],
    resourceName: '[RM] Web App',
  },
  FUNCTION_APP: {
    _type: 'azure_function_app',
    _class: ['Function'],
    resourceName: '[RM] Function App',
  },
};

export const AppServiceRelationships = {
  RESOURCE_GROUP_HAS_WEB_APP: createResourceGroupResourceRelationshipMetadata(
    AppServiceEntities.WEB_APP._type,
  ),
  RESOURCE_GROUP_HAS_FUNCTION_APP: createResourceGroupResourceRelationshipMetadata(
    AppServiceEntities.FUNCTION_APP._type,
  ),
};
