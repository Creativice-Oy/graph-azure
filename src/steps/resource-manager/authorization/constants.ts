import { PrincipalType } from '@azure/arm-authorization/esm/models';
import {
  GROUP_ENTITY_TYPE,
  STEP_AD_GROUPS,
  USER_ENTITY_TYPE,
  STEP_AD_USERS,
  SERVICE_PRINCIPAL_ENTITY_TYPE,
  STEP_AD_SERVICE_PRINCIPALS,
} from '../../active-directory';
import { generateRelationshipType } from '@jupiterone/integration-sdk-core';

// Fetch Role Assignments & Definitions
export const STEP_RM_AUTHORIZATION_ROLE_ASSIGNMENTS_AND_DEFINITIONS =
  'rm-authorization-role-assignments-and-definitions';

export const ROLE_DEFINITION_ENTITY_TYPE = 'azure_role_definition';
export const ROLE_DEFINITION_ENTITY_CLASS = ['AccessRole', 'AccessPolicy'];

export const ROLE_ASSIGNMENT_RELATIONSHIP_CLASS = 'ASSIGNED';

interface RoleAssignmentMap {
  principalType: PrincipalType;
  type: string;
  dependsOn: string[];
}

export const ROLE_ASSIGNMENT_TYPES_MAP: RoleAssignmentMap[] = [
  {
    principalType: 'Application',
    type: 'azure_application',
    dependsOn: [],
  },
  {
    principalType: 'DirectoryObjectOrGroup',
    type: 'azure_directory',
    dependsOn: [],
  },
  {
    principalType: 'DirectoryRoleTemplate',
    type: 'azure_directory_role_template',
    dependsOn: [],
  },
  {
    principalType: 'Everyone',
    type: 'azure_everyone',
    dependsOn: [],
  },
  {
    principalType: 'ForeignGroup',
    type: 'azure_foreign_group',
    dependsOn: [],
  },
  {
    principalType: 'Group',
    type: GROUP_ENTITY_TYPE,
    dependsOn: [STEP_AD_GROUPS],
  },
  {
    principalType: 'MSI',
    type: 'azure_msi',
    dependsOn: [],
  },
  {
    principalType: 'ServicePrincipal',
    type: SERVICE_PRINCIPAL_ENTITY_TYPE,
    dependsOn: [STEP_AD_SERVICE_PRINCIPALS],
  },
  {
    principalType: 'Unknown',
    type: 'azure_unknown',
    dependsOn: [],
  },
  {
    principalType: 'User',
    type: USER_ENTITY_TYPE,
    dependsOn: [STEP_AD_USERS],
  },
];

export const ROLE_ASSIGNMENT_DEFAULT_TYPE = 'azure_unknown_role_target';

export const ROLE_ASSIGNMENT_DEPENDS_ON = ([] as string[]).concat(
  ...ROLE_ASSIGNMENT_TYPES_MAP.map((t) => t.dependsOn),
);
export const ROLE_ASSIGNMENT_TARGET_ENTITY_TYPES = ([] as string[]).concat(
  ...ROLE_ASSIGNMENT_TYPES_MAP.map((t) => t.type),
);

export function getJupiterTypeForPrincipalType(
  principalType: PrincipalType | undefined,
): string {
  return (
    ROLE_ASSIGNMENT_TYPES_MAP.find((t) => t.principalType === principalType)
      ?.type || ROLE_ASSIGNMENT_DEFAULT_TYPE
  );
}

export function createRoleAssignmentRelationshipType(
  targetEntityType: string,
): string {
  return generateRelationshipType(
    ROLE_ASSIGNMENT_RELATIONSHIP_CLASS,
    ROLE_DEFINITION_ENTITY_TYPE,
    targetEntityType,
  );
}

export const ROLE_ASSIGNMENT_RELATIONSHIP_TYPES = [
  createRoleAssignmentRelationshipType(ROLE_ASSIGNMENT_DEFAULT_TYPE),
  ...ROLE_ASSIGNMENT_TARGET_ENTITY_TYPES.map(
    createRoleAssignmentRelationshipType,
  ),
];

// Fetch Classic Administrators
export const STEP_RM_AUTHORIZATION_CLASSIC_ADMINISTRATORS =
  'rm-authorization-classic-administrators';

export const CLASSIC_ADMINISTRATOR_ENTITY_KEY = 'azure_classic_admin_group';
export const CLASSIC_ADMINISTRATOR_ENTITY_TYPE = 'azure_classic_admin_group';
export const CLASSIC_ADMINISTRATOR_RELATIONSHIP_TYPE =
  'azure_classic_admin_group_has_user';
export const CLASSIC_ADMINISTRATOR_ENTITY_CLASS = 'UserGroup';
