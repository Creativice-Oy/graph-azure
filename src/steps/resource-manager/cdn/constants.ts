import {
  generateRelationshipType,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { createResourceGroupResourceRelationshipMetadata } from '../utils/createResourceGroupResourceRelationship';

export const STEP_RM_CDN_PROFILE = 'rm-cdn-profiles';
export const STEP_RM_CDN_ENDPOINTS = 'rm-cdn-endpoints';

export const CdnEntities = {
  PROFILE: {
    _type: 'azure_cdn_profile',
    _class: ['Service'],
    resourceName: '[RM] CDN Profile',
  },
  ENDPOINT: {
    _type: 'azure_cdn_endpoint',
    _class: ['Gateway'],
    resourceName: '[RM] CDN Endpoint',
  },
};

export const CdnRelationships = {
  RESOURCE_GROUP_HAS_PROFILE: createResourceGroupResourceRelationshipMetadata(
    CdnEntities.PROFILE._type,
  ),
  PROFILE_HAS_ENDPOINT: {
    _type: generateRelationshipType(
      RelationshipClass.HAS,
      CdnEntities.PROFILE,
      CdnEntities.ENDPOINT,
    ),
    sourceType: CdnEntities.PROFILE._type,
    _class: RelationshipClass.HAS,
    targetType: CdnEntities.ENDPOINT._type,
  },
};
