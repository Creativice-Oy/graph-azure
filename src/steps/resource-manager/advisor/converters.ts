import {
  Entity,
  createIntegrationEntity,
  convertProperties,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { AdvisorEntities } from './constants';
import { ResourceRecommendationBase } from '@azure/arm-advisor/esm/models';

function maybeConvertToNumber(maybeNumber: string | undefined) {
  if (!isNaN(parseFloat(maybeNumber || 'NaN'))) return parseFloat(maybeNumber!);
}

export function createRecommendationEntity(
  webLinker: AzureWebLinker,
  data: ResourceRecommendationBase,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: AdvisorEntities.RECOMMENDATION._type,
        _class: AdvisorEntities.RECOMMENDATION._class,
        id: data.id,
        name: data.name,
        category: data.category,
        impact: data.impact,
        impactedField: data.impactedField,
        impactedValue: data.impactedValue,
        recommendationTypeId: data.recommendationTypeId,
        shortDescriptionProblem: data.shortDescription?.problem,
        shortDescriptionSolution: data.shortDescription?.solution,
        assessmentKey: data.extendedProperties?.assessmentKey,
        score: maybeConvertToNumber(data.extendedProperties?.score),
        resourceId: data.resourceMetadata?.resourceId,
        source: data.resourceMetadata?.source,
        open: true,
        severity: data.impact,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
