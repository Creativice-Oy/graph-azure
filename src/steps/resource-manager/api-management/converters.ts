import {
  Entity,
  createIntegrationEntity,
  convertProperties,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { ApiManagementEntities } from './constants';
import {
  ApiManagementServiceResource,
  ApiContract,
} from '@azure/arm-apimanagement/esm/models';

export function createApiManagementServiceEntity(
  webLinker: AzureWebLinker,
  data: ApiManagementServiceResource,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: ApiManagementEntities.SERVICE._type,
        _class: ApiManagementEntities.SERVICE._class,
        id: data.id,
        name: data.name,
        public: !!data.publicIPAddresses?.length,
        function: ['api-gateway'],
        category: ['application'],
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}

export function createApiManagementApiEntity(
  webLinker: AzureWebLinker,
  data: ApiContract,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: ApiManagementEntities.API._type,
        _class: ApiManagementEntities.API._class,
        id: data.id,
        name: data.name,
        address: data.path,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
