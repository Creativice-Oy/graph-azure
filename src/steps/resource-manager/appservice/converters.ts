import {
  Entity,
  createIntegrationEntity,
  StepEntityMetadata,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { AppServicePlan, Site } from '@azure/arm-appservice/esm/models';
import { AppServiceEntities } from './constants';

export function createAppEntity(
  webLinker: AzureWebLinker,
  data: Site,
  metadata: StepEntityMetadata,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: data.id as string,
        _type: metadata._type,
        _class: metadata._class,
        id: data.id,
        name: data.name,
        type: data.type,
        kind: data.kind?.split(','),
        location: data.location,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}

export function createAppServicePlanEntity(
  webLinker: AzureWebLinker,
  data: AppServicePlan,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: data.id as string,
        _type: AppServiceEntities.APP_SERVICE_PLAN._type,
        _class: AppServiceEntities.APP_SERVICE_PLAN._class,
        id: data.id,
        name: data.name,
        type: data.type,
        kind: data.kind?.split(','),
        location: data.location,
        'sku.name': data.sku?.name,
        'sku.tier': data.sku?.tier,
        'sku.size': data.sku?.size,
        'sku.family': data.sku?.family,
        'sku.capacity': data.sku?.capacity,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
