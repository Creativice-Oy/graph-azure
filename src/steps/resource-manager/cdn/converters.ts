import {
  Entity,
  createIntegrationEntity,
  convertProperties,
} from '@jupiterone/integration-sdk-core';

import { AzureWebLinker } from '../../../azure';
import { CdnEntities } from './constants';
import { Profile, Endpoint } from '@azure/arm-cdn/esm/models';

export function createCdnProfileEntity(
  webLinker: AzureWebLinker,
  data: Profile,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: CdnEntities.PROFILE._type,
        _class: CdnEntities.PROFILE._class,
        id: data.id,
        name: data.name,
        category: ['infrastructure'],
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}

export function createCdnEndpointEntity(
  webLinker: AzureWebLinker,
  data: Endpoint,
): Entity {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        ...convertProperties(data),
        _key: data.id as string,
        _type: CdnEntities.ENDPOINT._type,
        _class: CdnEntities.ENDPOINT._class,
        id: data.id,
        name: data.name,
        category: ['data'],
        function: ['content-distribution'],
        public: true,
        webLink: webLinker.portalResourceUrl(data.id),
      },
    },
  });
}
