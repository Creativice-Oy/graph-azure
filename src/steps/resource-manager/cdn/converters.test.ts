import { createAzureWebLinker } from '../../../azure';
import { createCdnProfileEntity, createCdnEndpointEntity } from './converters';
import { Profile, Endpoint } from '@azure/arm-cdn/esm/models';
import { CdnEntities } from './constants';

const webLinker = createAzureWebLinker('something.onmicrosoft.com');

describe('createContainerRegistryEntity', () => {
  test('properties transferred', () => {
    const data: Profile = {
      id:
        '/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/Microsoft.Cdn/profiles/j1dev',
      name: 'j1dev',
      type: 'Microsoft.Cdn/profiles',
      location: 'EastUs',
      tags: {},
      sku: { name: 'Standard_Microsoft' },
      resourceState: 'Active',
      provisioningState: 'Succeeded',
    };

    expect(createCdnProfileEntity(webLinker, data)).toMatchSnapshot();
    expect(createCdnProfileEntity(webLinker, data)).toMatchGraphObjectSchema({
      _class: CdnEntities.PROFILE._class,
      schema: {},
    });
  });
});

describe('createContainerRegistryWebhookEntity', () => {
  test('properties transferred', () => {
    // the following is returned from the API, but current SDK version doesn't recognize `systemData` as a property on `Webhook`
    const data: Endpoint = {
      id:
        '/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/Microsoft.Cdn/profiles/j1dev/endpoints/j1dev',
      name: 'j1dev',
      type: 'Microsoft.Cdn/profiles/endpoints',
      location: 'EastUs',
      tags: {},
      originPath: undefined,
      contentTypesToCompress: [],
      originHostHeader: 'www.jupiterone.com',
      isCompressionEnabled: false,
      isHttpAllowed: true,
      isHttpsAllowed: true,
      queryStringCachingBehavior: 'IgnoreQueryString',
      optimizationType: undefined,
      probePath: undefined,
      geoFilters: [],
      defaultOriginGroup: undefined,
      urlSigningKeys: [],
      deliveryPolicy: { description: '', rules: [] },
      webApplicationFirewallPolicyLink: undefined,
      hostName: 'j1dev.azureedge.net',
      origins: [
        {
          name: 'j1dev',
          hostName: 'www.jupiterone.com',
          httpPort: 80,
          httpsPort: 443,
          originHostHeader: 'www.jupiterone.com',
          priority: 1,
          weight: 1000,
          enabled: true,
          privateLinkAlias: undefined,
          privateLinkResourceId: undefined,
          privateLinkLocation: undefined,
          privateLinkApprovalMessage: undefined,
        },
      ],
      originGroups: [],
      resourceState: 'Running',
      provisioningState: 'Succeeded',
    };

    expect(createCdnEndpointEntity(webLinker, data)).toMatchSnapshot();
    expect(createCdnEndpointEntity(webLinker, data)).toMatchGraphObjectSchema({
      _class: CdnEntities.ENDPOINT._class,
      schema: {},
    });
  });
});
