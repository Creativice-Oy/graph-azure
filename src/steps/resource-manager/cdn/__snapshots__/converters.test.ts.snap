// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createContainerRegistryEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Service",
  ],
  "_key": "/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/Microsoft.Cdn/profiles/j1dev",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "id": "/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/Microsoft.Cdn/profiles/j1dev",
        "location": "EastUs",
        "name": "j1dev",
        "provisioningState": "Succeeded",
        "resourceState": "Active",
        "sku": Object {
          "name": "Standard_Microsoft",
        },
        "tags": Object {},
        "type": "Microsoft.Cdn/profiles",
      },
    },
  ],
  "_type": "azure_cdn_profile",
  "category": Array [
    "infrastructure",
  ],
  "createdOn": undefined,
  "displayName": "j1dev",
  "id": "/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/Microsoft.Cdn/profiles/j1dev",
  "location": "EastUs",
  "name": "j1dev",
  "provisioningState": "Succeeded",
  "resourceState": "Active",
  "type": "Microsoft.Cdn/profiles",
  "webLink": "https://portal.azure.com/#@something.onmicrosoft.com/resource/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/Microsoft.Cdn/profiles/j1dev",
}
`;

exports[`createContainerRegistryWebhookEntity properties transferred 1`] = `
Object {
  "_class": Array [
    "Gateway",
  ],
  "_key": "/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/Microsoft.Cdn/profiles/j1dev/endpoints/j1dev",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "contentTypesToCompress": Array [],
        "defaultOriginGroup": undefined,
        "deliveryPolicy": Object {
          "description": "",
          "rules": Array [],
        },
        "geoFilters": Array [],
        "hostName": "j1dev.azureedge.net",
        "id": "/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/Microsoft.Cdn/profiles/j1dev/endpoints/j1dev",
        "isCompressionEnabled": false,
        "isHttpAllowed": true,
        "isHttpsAllowed": true,
        "location": "EastUs",
        "name": "j1dev",
        "optimizationType": undefined,
        "originGroups": Array [],
        "originHostHeader": "www.jupiterone.com",
        "originPath": undefined,
        "origins": Array [
          Object {
            "enabled": true,
            "hostName": "www.jupiterone.com",
            "httpPort": 80,
            "httpsPort": 443,
            "name": "j1dev",
            "originHostHeader": "www.jupiterone.com",
            "priority": 1,
            "privateLinkAlias": undefined,
            "privateLinkApprovalMessage": undefined,
            "privateLinkLocation": undefined,
            "privateLinkResourceId": undefined,
            "weight": 1000,
          },
        ],
        "probePath": undefined,
        "provisioningState": "Succeeded",
        "queryStringCachingBehavior": "IgnoreQueryString",
        "resourceState": "Running",
        "tags": Object {},
        "type": "Microsoft.Cdn/profiles/endpoints",
        "urlSigningKeys": Array [],
        "webApplicationFirewallPolicyLink": undefined,
      },
    },
  ],
  "_type": "azure_cdn_endpoint",
  "category": Array [
    "data",
  ],
  "createdOn": undefined,
  "displayName": "j1dev",
  "function": Array [
    "content-distribution",
  ],
  "hostName": "j1dev.azureedge.net",
  "id": "/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/Microsoft.Cdn/profiles/j1dev/endpoints/j1dev",
  "isCompressionEnabled": false,
  "isHttpAllowed": true,
  "isHttpsAllowed": true,
  "location": "EastUs",
  "name": "j1dev",
  "originHostHeader": "www.jupiterone.com",
  "provisioningState": "Succeeded",
  "public": true,
  "queryStringCachingBehavior": "IgnoreQueryString",
  "resourceState": "Running",
  "type": "Microsoft.Cdn/profiles/endpoints",
  "webLink": "https://portal.azure.com/#@something.onmicrosoft.com/resource/subscriptions/d3803fd6-2ba4-4286-80aa-f3d613ad59a7/resourcegroups/j1dev/providers/Microsoft.Cdn/profiles/j1dev/endpoints/j1dev",
}
`;
